### PR TITLE
resnet v1 in fp16

### PR DESCRIPTION
**Pull request by ap-hynninen**
Implements the original resnet network in fp16 precision

### CODE REVIEW

File: example/image-classification/symbols/resnet-v1-fp16.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    # Although kernel is not used here when global_pool=True, we should put one
+    pool1 = mx.symbol.Pooling(data=body, global_pool=True, kernel=(7, 7), pool_type='avg', name='pool1')
+    flat = mx.symbol.Flatten(data=pool1)
+    weight = mx.symbol.Variable(name='fc1_weight', dtype=np.float32)
+    bias = mx.symbol.Variable(name='fc1_bias', dtype=np.float32)
+    weight = mx.symbol.Cast(data=weight, dtype=np.float16)
+    bias = mx.symbol.Cast(data=bias, dtype=np.float16)
+    fc1 = mx.symbol.FullyConnected(data=flat, weight=weight, bias=bias, num_hidden=num_classes, name='fc1')
+    label = mx.symbol.Variable(name='softmax_label')
+    label = mx.symbol.Cast(data=label, dtype=np.float16)
```
Comment (piiswrong) [inline example/image-classification/symbols/resnet-v1-fp16.py]: casting label to fp16 might be a problem...

Comment (ap-hynninen) [inline example/image-classification/symbols/resnet-v1-fp16.py]: Hmm.. I thought so too, but for some reason I thought I needed that. Maybe I don't?

Comment (piiswrong) [inline example/image-classification/symbols/resnet-v1-fp16.py]: you can cast output to fp32 and do softmax in fp32 instead

**Comment by leonid-pishchulin** [comment]
I’m trying to run MXnet resnet-18 model on 640x480 images using resnet_fp16.py
on Jetson TX1, but I observe only very little performance difference compared to regular resnet.py (103 ms/fr vs. 111 ms/fr found by averaging over 1000 runs)
 
Do I need to build mxnet specifically with fp16 support or smth like that? I use CUDA 8.0 and cuDNN 5.1
 
Do I need to cast activations to fp16?
act1 = mx.sym.Activation(data=bn1, act_type='relu', name=name + '_relu1')
act1 = mx.symbol.Cast(data=act1, dtype=np.float16)

**Comment by ap-hynninen** [comment]
The problem with fp16 is that there's a lot of overhead currently on fp32->fp16 conversions which eats into the performance gain you get from faster GEMMs. Have you profiled the run with nvprof/nvvp? It would be interesting to know if there are other bottlenecks.

On top of my head, the activations should be done in fp16 already since the input is in fp16.

**Comment by ap-hynninen** [comment]
On Jetson TX1 you may be CPU limited. You can check for this by running via profiler and seeing if there are gaps in the CUDA kernel execution

**Comment by leonid-pishchulin** [comment]
I benchmarked performance on TX2, and didn't see any significant difference between fp16 and fp32 either.
I built mxnet on TX2 with USE_PASCAL=1 flag and run profiling of resnet 18 example with fp16 and fp32 available from the repository files resnet.py, resnet_fp16.py, respectively.
I use batch size of 20 images and average the results over 50 times. Prior to run with the profiler, I perform a dry run with the same batch size of 20 and number of runs = 5. For profiling I follow the example https://github.com/dmlc/mxnet/blob/master/example/profiler/profiler_executor.py

I get 15.24 ms/frame​ for fp16 and 15.67 ms/frame for fp32
I attach a screen shot of the profiler output for one of the batches in each case.
profiler-resnet-18-fp16-tx2

**Comment by ptrendx** [comment]
First of all, could you try with batch size being a power of 2? This should help in general.
Secondly, could you run the code through nvprof and post somewhere the result? It should look something like this:
`nvprof -o profile.nvvp ...`
where `...` is the command you normally run to launch the test. This should show us whether the correct kernels are being launched. and likely give some more insight into why there is no improvement.

**Comment by leonid-pishchulin** [comment]
I run nvprof with resnet.py and resnet_fp16.py with batch size 1 and # runs = 1000 (640x480 images) on Jetson TX1.
resnet.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBUGdhQVZXYnFTMjg
resnet_fp16.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBSmNJSkdjZXByZ2M

overall it seems that in case of resnet_fp16.py *_fp16_* kernels are called, i.e. maxwell_fp16_scudnn_fp16_128x128_small_nn vs. maxwell_scudnn_128x128_small_nn in case of resnet.py, which is a good sign. However, maxwell_fp16_scudnn_fp16_128x128_small_nn kernel takes 63.5 s in total for all 1000 runs, whereas maxwell_scudnn_128x128_small_nn takes 44.3 s in total.

increasing batch size to 2 results into 28.6 s for maxwell_fp16_scudnn_fp16_128x128_small_nn vs. 39.9 s for maxwell_scudnn_128x128_small_nn, which is encouraging. However, the second most time consuming kernel maxwell_fp16_scudnn_winograd_fp16_128x128_mobile_tile148t_nt takes 21.7 s vs. 14.7 s in case of maxwell_scudnn_winograd_128x128_mobile_tile148t_nt. Thus overall I don't observe any speed-up due to switching to fp16. 

Any ideas?

**Comment by leonid-pishchulin** [comment]
batch=16:
resnet.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBczNCRnkwTjI3SlU
resnet_fp16.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBdXZMNHMzRFI1aG8

maxwell_fp16_scudnn_winograd_fp16_128x128_mobile_tile148t_nt: 13.9 s
maxwell_scudnn_winograd_128x128_mobile_tile148t_nt: 15.9 s

maxwell_fp16_scudnn_fp16_128x64_medium_nn: 6.3 s
maxwell_scudnn_128x64_medium_nn: 6.4 s

maxwell_fp16_scudnn_fp16_128x64_small_nn:​ 5.8
maxwell_scudnn_128x64_small_nn​: 6.0 s

all kernels were invoked the same number of times

**Comment by ptrendx** [comment]
So, 2 things:
 - existence of winograd kernels means that you do training - are you sure you want to do that vs just inference on Tegra?
 - the calls you see in the fp16 case are pseudo-fp16 calls (input/output is in fp16, but the math is actually done in fp32), so it is understandable that you get similar perf for them. I was under impression that setting MSHADOW_USE_PASCAL flag during compilation should do calls to true-fp16 versions. Are you sure you set that flag?

**Comment by leonid-pishchulin** [comment]
this particular results were obtained using TX1. MSHADOW_USE_PASCAL=0 in this case, as I assumed this has to be set to 1 only for TX2.
this is how I run the mxnet
sym = resnet.get_symbol(num_classes=10,num_layers=18,image_shape=image_shape)
mod = mx.mod.Module(symbol=sym, context=mx.gpu(device_id))
mod.bind(for_training=False, inputs_need_grad=False, data_shapes=[('data', (batch_size,3,Y,X))])
...
d = mx.nd.array(d,mx.gpu(device_id))
Batch = namedtuple('Batch', ['data'])
mod.forward(Batch([d]))

running the same way on TX2 compiled with MSHADOW_USE_PASCAL=1 shows no difference either

**Comment by ptrendx** [comment]
I retract my comment about training - I looked at the profiles and you are doing only inference in those runs.

"running the same way" means "it does not give a speed difference too" or "I profiled it and also see pseudofp16 versions of kernels"? Please do the profile.

Adding @DickJC123, since he was experimenting a lot with pseudo- vs true-fp16 paths in MXNet. I assumed that setting MSHADOW_USE_PASCAL should give true-fp16 kernels. Is that not true?

**Comment by DickJC123** [comment]
On the TX2, you should compile MXNET with MSHADOW_USE_PASCAL=1 to get true fp16 kernels.  Sounds like you're doing that though.  Two possibilities then to help:

- Can you upgrade to cuDNNv6? MXNET with cuDNNv5.1 uses a different call (compared to v6) to set the convolution descriptor (one that does not explicitly set the compute precision).  Try this first.
- Can you run your experiment temporarily on a beefier PASCAL brick to check for a different behavior? There are subtle differences in the resources available to kernels run on a TX2 vs. say a TitanX-PASCAL. See the table at the bottom of https://en.wikipedia.org/wiki/CUDA.  I don't know if all algos are prepared to run on architecture sm_62.

And although this shouldn't affect your problem, you should add "62" to KNOWN_CUDA_ARCHS in ./Makefile (available in a recent push).  Remove other archs you don't care about to speed mxnet lib builds.

**Comment by leonid-pishchulin** [comment]
@ptrendx : I profiled on TX2 with cuDNN 5.1, CUDA 8.0, MSHADOW_USE_PASCAL = 1, batch_size = 4
resnet.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBZERUVWRfRzFpa0E
resnet_fp16.py profile output: https://drive.google.com/open?id=0Bwb8o6lldJgBRDI4OUpveVJZckk

maxwell_fp16_scudnn_winograd_fp16_128x128_mobile_tile148t_nt: 31.9 s / 3661 invocations = 8.7 ms/invocation
maxwell_scudnn_winograd_128x128_mobile_tile148t_nt: 22.8 s / 2818 invocations = 8 ms/invocation

next step: test with cuDNN 6.0

**Comment by ptrendx** [comment]
Yes, do test with cudnn 6. Now that Dick mentioned it I remember that without it you can't get true fp16 even with use_pascal flag set.

**Comment by leonid-pishchulin** [comment]
I tested with cuDNN 6 when using JetPack 3.1 on TX2. Profiling of resnet_fp16.py shows that true FP16 cuda kernels are called (*_hcudnn_*) and no more *_fp16_* calls. Following findings have been made while profiling resnet_fp16.py with number layers = 18 and with batch size 8:
- FP16 convolutional kernels are 1.3-1.8x faster compared to FP32
- FP16 batch normalization kernels are 1.2-1.4x faster compared to FP32
- FP16 activation (RELU) kernels are also 1.2-1.4x faster compared to FP32
- overall cost of casting FP32 to FP16 is only 2% of the total run-time
Is that to be expected?
Profiler output is here: https://drive.google.com/open?id=0Bwb8o6lldJgBSUh0U2NEZXVvRW8
Overall, I observe 1.5x speed-up due to FP16 for batch size of 8 and input image resolution of 640x480, and increasing batch results into the similar speed-ups of 1.5x.
I have a few questions:
- reducing batch size to 4 and 1 shows much smaller gains of 1.2x and 1.1x due to FP16, respectively. Is that to be expected? Presumably this is due to the overhead of copying the data to/from GPU.
- keeping batch size = 8, but reducing by 2x the number of filters per layer shows a speed-up FP16 vs. FP32 of only 1.3x. Why is that?
-  reducing input image resolution to 0.7*original resolution does not affect significantly the gains due to FP16 for different batch sizes, compared to using original resolution
- when profiling my custom FP16 ResNet-18 that has dilated convolutions, I see pseudo-FP16 calls *_fp16_* for the dilated convolutions, and thus no speed-up over the corresponding FP32 kernels. Do you know if there is a timeline of supporting dilated convolutions?
- FP16 deconvolutional kernels cudnn::detail::dgrad2d_alg1_1_half<__half...> perform 10-15% worse compared to the corresponding FP32 deconvolutional kernels. Why?
Thanks a lot for your help !
